### PR TITLE
[autograd] remove spinning for dist engine

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36606 [autograd] remove spinning for dist engine**
* #37061 [autograd] refactor some functions

This PR refactor the continuation logic of the async mode on autograd
engine, to avoid launch spinning works. To achieve that:
1. remove the continuation logic in
execute_graph_task_with_continuiation
2. separate the usage of execute_graph_task between dist_engine and
local engine, now dist_engine universally use
`execute_graph_task_until_ready_queue_empty` (a better name appreciated
here).
3. remove enqueue_blocked_task_on_cpu
4. remove the async mode in `execute_with_graph_task` as we don't need
to use it in dist_engine
5. move some Engine functions to public to make it accessible from dist
engine

Differential Revision: [D21032731](https://our.internmc.facebook.com/intern/diff/D21032731)

### CODE REVIEW

File: torch/csrc/autograd/engine.cpp [hunk 1]
```diff
@@ -809,35 +809,35 @@ void Engine::initialize_device_threads_pool() {
...
-    }
     {
+      NodeTask task = graph_task_rq->pop();
+      if (!(local_graph_task = task.base_.lock())) {
+        continue;
+      }
       if (task.fn_ && !local_graph_task->has_error_.load()) {
         AutoGradMode grad_mode(local_graph_task->grad_mode_);
         try {
           evaluate_function(local_graph_task, task.fn_.get(), task.inputs_);
```
Comment (gqchen) [inline torch/csrc/autograd/engine.cpp:834]: we need to do continuation in evaluate_function? It seems it's not part of this PR.

File: torch/csrc/autograd/engine.cpp [hunk 2]
```diff
@@ -809,35 +809,35 @@ void Engine::initialize_device_threads_pool() {
...
-  // The graph_task must be alive at this point, because internal autograd machinary
-  // who calls this API (Distributed Autograd Engine) increases outstanding_tasks_
-  // outside this API to keep the GraphTask alive
-  TORCH_INTERNAL_ASSERT(graph_task, "GraphTask is no longer valid!");
-  ready_queue(graph_task, at::kCPU)->push(
-      std::move(task), /* incrementOutstandingTasks */ false);
-}
-
-void Engine::execute_graph_task_with_continuation(const std::shared_ptr<GraphTask>& graph_task) {
   std::shared_ptr<ReadyQueue> graph_task_rq = graph_task->cpu_ready_queue_;
```
Comment (gqchen) [inline torch/csrc/autograd/engine.cpp:817]: If we share one ready queue for the whole graph task, everytime a new backward is called, we create a new worker which can take nodes from the same readyqueue other backward calls populated. Our general threading strategy is fork and rpc in the forward path makes the corresponding backward  parallelizable, and ops in the corresponding "fork" and "rpc" should be run in the same thread. The behavior here makes backward and forward unsymmetric. It will be much harder to reason and explain.

File: torch/csrc/autograd/engine.cpp [hunk 3]
```diff
@@ -850,19 +850,13 @@ void Engine::execute_graph_task_with_continuation(const std::shared_ptr<GraphTas
...
     // would notify the owner thread that the task has been completed.
     mark_graph_task_completed(graph_task);
-  } else {
-    // schedule a continuation
-    at::launch([this, graph_task]() {
-        execute_graph_task_with_continuation(graph_task);
-    });
   }
-
+  return graph_task->future_result_;
```
Comment (gqchen) [inline torch/csrc/autograd/engine.cpp:854]: this PR doesn't address refcount issue? Without that fix, this will cause training failures.

**Review by gqchen — CHANGES_REQUESTED** [review]

File: torch/csrc/autograd/engine.cpp (hunk 3 above)
Comment (wanchaol) [reply torch/csrc/autograd/engine.cpp:854]: it seems that we don't need to change to return `shared_ptr<FutureVariableList>&` as of @jjlilley 's change in https://github.com/pytorch/pytorch/pull/36502, I will need to verify if this is the case.

File: torch/csrc/autograd/engine.cpp (hunk 1 above)
Comment (gqchen) [reply torch/csrc/autograd/engine.cpp:834]: We certainly don't want to change local engine. 

I think what we want to avoid here is to get stuck silently when someone happens to do distributed autograd using GPU. Can we add an assertion here if node is expected to run on GPU?

File: torch/csrc/autograd/engine.cpp (hunk 2 above)
Comment (gqchen) [reply torch/csrc/autograd/engine.cpp:817]: definitely.

**Comment by wanchaol** [comment]
@gqchen I just had a offline discussion with @albanD, we will refactor the existing engine `thread_main` to make it works well with both the vanilla engine and dist engine. So that in dist_engine, we don't need to increase `outstanding_task` to keep the graph_task alive, and could remove this `execute_graph_task_until_ready_queue_empty` redundant logic. This would serve us better on solving GPU training issue (no continuation needed, we will just waiting for the GPU task to finish before we exit the thread_main).

**Comment by gqchen** [comment]
That's great!

**Comment by wanchaol** [comment]
@albanD this is now ready for review.

File: torch/csrc/distributed/autograd/engine/dist_engine.cpp [hunk 4]
```diff
@@ -254,17 +248,71 @@ void DistEngine::computeDependencies(
...
+      // block and can be deallocated (release any references to grad tensors
+      // as part of inputs_)
+      NodeTask task = cpu_ready_queue->pop();
+      if (!(local_graph_task = task.base_.lock())) {
+        continue;
+      }
+      if (task.fn_ && !local_graph_task->has_error_.load()) {
+        AutoGradMode grad_mode(local_graph_task->grad_mode_);
+        try {
+          engine_.evaluate_function(local_graph_task, task.fn_.get(), task.inputs_, cpu_ready_queue);
```
Comment (albanD) [inline torch/csrc/distributed/autograd/engine/dist_engine.cpp:278]: Not sure if you this is supposed to support reentrant autograd. If so, I think it is missing an `init_thread_local_ready_queue()`. Otherwise, it's fine.

**Review by albanD — APPROVED** [review]
The vanilla engine part looks good.

Small nit in the distributed one if you want to support reentrant. Otherwise it looks ok from a high level.

File: torch/csrc/distributed/autograd/engine/dist_engine.cpp (hunk 4 above)
Comment (wanchaol) [reply torch/csrc/distributed/autograd/engine/dist_engine.cpp:278]: I think in the dist engine call phase, it's a bit hard now to support reentrant call from `distributed::backward()` level, but for local reentrant backward that dist backward calls, it might be possible to support it. The tricky here is, we are using a locally created ready_queue per `execute_until_ready_queue_empty`, to execute the same graph task concurrently, if we `init_thread_local_ready_queue()`, there's chance that the thread local ready_queue get overwrited by two different threads concurrently (e.g. thread A and B doing local reentrant backward concurrently). We might need to think more on how to handle local reentrant backward calls indeed, right now the approach is to use a new ready queue whenever local reentrant calls `Engine::execute`. I would leave it as a follow up task.

File: torch/csrc/distributed/autograd/engine/dist_engine.h [hunk 5]
```diff
@@ -88,13 +88,47 @@ class TORCH_API DistEngine {
...
+  // When `incrementOutstandingTasks=false`, the function does not increment 
+  // 'outstanding_tasks_' in the appropriate GraphTask. It is assumed we've already
+  // done this before hand for this task (to ensure we don't pre-mark this graph_task
+  // as completed). This is useful in the distributed autograd case where we need to
+  // increment 'outstanding_tasks_' first to indicate the local autograd engine the
+  // graph task is not completed until it receives the signals from other workers
+  // over the network.
+  //
+  // XXX: calling this function assumes that we will have NO GPU nodetasks be executed
+  // for the graph_task, the caller of this function need to ensure this otherwise
```
Comment (gqchen) [inline torch/csrc/distributed/autograd/engine/dist_engine.h:111]: can we do a follow up task to add some assertion for this?

File: torch/csrc/distributed/autograd/engine/dist_engine.cpp [hunk 6]
```diff
@@ -254,17 +248,71 @@ void DistEngine::computeDependencies(
   autogradContext->setGraphTask(std::move(graphTask));
 }
 
+std::shared_ptr<FutureVariableList>& DistEngine::execute_graph_task_until_ready_queue_empty(
```
Comment (gqchen) [inline torch/csrc/distributed/autograd/engine/dist_engine.cpp:251]: love the idea of moving this function to DistEngine!

**Review by gqchen — APPROVED** [review]
looks great!

File: torch/csrc/distributed/autograd/engine/dist_engine.cpp (hunk 4 above)
Comment (albanD) [reply torch/csrc/distributed/autograd/engine/dist_engine.cpp:278]: But this is a thread local variable. so different thread should not overwrite each other right?

I am fine if this is not supported, just wanted to point it out. Maybe you want to add a comment to make sure this is clear.

Comment (wanchaol) [reply torch/csrc/distributed/autograd/engine/dist_engine.cpp:278]: Oh actually you are right, since it's a thread local it will be preserved correctly. Let me try to add it :)

File: torch/csrc/autograd/engine.cpp [hunk 7]
```diff
@@ -1032,7 +974,6 @@ bool Engine::is_checkpoint_valid() {
 void Engine::init_local_ready_queue(std::shared_ptr<ReadyQueue> ready_queue) {
   if (ready_queue) {
     // if ready_queue provided in the caller, use the caller's ready_queue to initialize local_ready_queue
-    TORCH_INTERNAL_ASSERT(!local_ready_queue);
```
Comment (albanD) [inline torch/csrc/autograd/engine.cpp:1035]: Why do you need to remove that? This should always be true, even from the distributed call no? Otherwise you're doing reentrant for a dist engine? :D

Comment (wanchaol) [reply torch/csrc/autograd/engine.cpp:1035]: In the distributed engine the `execute_graph_task_until_rq_empty` is called from a thread pool (RPC threadpool or JIT threadpool), so if the thread pool launches the function twice or more, the local_ready_queue already been set to the ready queue that's created by previous calls. That's why I remove the assert here.

Comment (albanD) [reply torch/csrc/autograd/engine.cpp:1035]: Ok,
I might be cleaning this all logic later. Will be fine for now.

Comment (wanchaol) [reply torch/csrc/autograd/engine.cpp:1039]: actually I did a couple of runs with initializing the local_ready_queue to support reentrant backwards, it turns out the perf becomes bad after that (around 10% drop), this is pretty surprising as I wouldn't expect a big drop like this. So I reverted this change and added a TODO in the comment to investigate further in the future.
